smithy4s/src/main/scala/com/dwolla/consul/smithy4s/ConsulDiscoverable.scala

```diff
@@ -0,0 +1,17 @@
+package com.dwolla.consul.smithy4s
+
+import org.http4s.Uri
+import org.http4s.Uri.Host
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Instances are only available for Smithy4s Services annotated with @discoverable")
+trait ConsulDiscoverable[Alg[_[_]]] {
```

Comment:
I'm open to bikeshedding on the names here. This typeclass will have macro-instances for Smithy4s services annotated with the `@discoverable` trait. The `host`, `uriAuthority`, and `uri` values will be derived from the `serviceName` value on the `@discoverable` trait, e.g. `uri` will return `consul://${serviceName}`.

This will typically be used like this:
```scala
def smithyClient[F[_], Alg[_[_, _, _, _, _]]](client: Client[F],  // http4s client
                                              consulMiddleware: ConsulMiddleware[F], // middleware from this library
                                              service: Service[Alg], // generated by smithy4s
                                             )
                                             (implicit C: ConsulDiscoverable[service.Impl]): Resource[F, service.Impl[F]] =
  SimpleRestJsonBuilder.apply(service)
    .client(NatchezMiddleware.clientWithAttributes(client)(
      "rpc.system" -> "smithy",
      "peer.service" -> ConsulDiscoverable[service.Impl].host,
    ))
    .uri(ConsulDiscoverable[service.Impl].uri)
    .middleware(consulMiddleware)
    .resource
```

Reply:
And just to be clear, this does _not_ work, because the macro can't resolve what `service` is at compile time:
```scala
def smithyClient[F[_], Alg[_[_, _, _, _, _]]](client: Client[F],  // http4s client
                                              consulMiddleware: ConsulMiddleware[F], // middleware from this library
                                              service: Service[Alg], // generated by smithy4s
                                             ): Resource[F, service.Impl[F]] =
  SimpleRestJsonBuilder.apply(service)
    .client(NatchezMiddleware.clientWithAttributes(client)(
      "rpc.system" -> "smithy",
      "peer.service" → HostFromService(service),
    ))
    .uri(UriFromService(service))
    .middleware(consulMiddleware)
    .resource
```

Implementing it as a typeclass does work, because there is a stable location (outside the function above) where the service is known and the `ConsulDiscoverable` instance can be generated and then passed implicitly into the function.